### PR TITLE
Add parameter-clipping strategy to routines that generate samples (fix #846)

### LINKED ISSUE #846: parameter sampling, soft and hard limits, and the sample*flux functions
So, `sample_energy_flux` creates `num` realizations of the free-parameter distribution (e.g. if the model has nh, gamma, and ampl free then you get a num by 3 array of numbers) which is used to calculate the flux distribution. Since the numbers are generated using a normal distribution (either with correlated parameters or not) then you can end up with parameter values outside the hard-limit range (often seen with the nh parameter), so the code goes through and makes sure the parameters are clipped to the hard limits (that is, values outside the hard-limit range get replaced with values at the hard limit).

The `sample_flux` function currently post-processes this array to remove any flux values with parameters outside the soft-limit range (internally `sample_flux` calls `sample_energy_flux` to do most of the work). Unfortunately there are issues with how this is done (e.g. #751).

I suggest that add an option to `sample_energy_flux` that lets you chose whether to clip the parameters by the soft or hard limit (this would leave in those rows where values are outside the limits but replaces the problem value with the limit). This would then avoid the problem in #751 (and avoids `sample_flux` doing excess work).

One question is that I have left the bins in (since they provide some information related to upper limits, e.g. #457). However, maybe you would want to remove the clipped bins. One thought I have is to add an extra column indicating whether the row was "okay" or "clipped"  (e.g. 1 or 0) so users could filter out the clipped values if they wanted.

### PR DESCRIPTION
**Pull request by DougBurke**
# Summary

The addition of the clip parameter lets users control how parameter values are clipped before use in  sample_energy_flux, sample_photon_flux, plot_energy_flux, plot_photon_flux, get_energy_flux_hist, and get_photon_flux_hist.

# Details

At present only sherpa.astro.flux.sample_flux supports the `clip` parameter. The sherpa.astro.flux.calc_sample_flux is waiting on #754 before it is worth updating (this means that we can not add a clip parameter to sample_flux yet).

### CODE REVIEW

**Comment by DougBurke** [comment]
Note to self: was about to ask for SDS review, but I think we need #803 merged first (otherwise it's confusing to work out what is new in this PR)

**Comment by DougBurke** [comment]
Due to changes to the master branch this needs rebasing (since it's based no #803 which has just been rebased), but I don't have the energy to fix this at the moment.

**Comment by DougBurke** [comment]
I've rebased this and marked it ready for review (by SDS).

**Comment by hamogu** [comment]
In the combined set there are > 3000 lines to look at and for most of the files, github does not even show me the diff because it's so big. I don't even know where to start. That's one problem with these huge changes and PRs that build on top of each other. Reviews just become difficult.

Can this wait until #803 is merged?

**Comment by DougBurke** [comment]
For @hamogu - I'm labelling this as SDS review so we all have it on our radars. 

If you were bored and wanted to look at what this code is doing from a high level you could look at the new `clip` parameter to `sample_energy_flux` and see if the documentation makes sense.

**Comment by DougBurke** [comment]
Rebased because of recent changes and to fix a less-than-ideal rebase.

**Comment by DougBurke** [comment]
Rebased since #803 has been merged.

**Comment by DougBurke** [comment]
So @anetasie and @hamogu - this should be easier to review now.

**Review by anetasie — APPROVED** [review]
It looks good. I agree with the default clip setting to the hard limits.

File: sherpa/astro/flux.py
Comment (wmclaugh) [inline sherpa/astro/flux.py:180]: paraneters -> parameters

File: sherpa/astro/flux.py
Comment (wmclaugh) [inline sherpa/astro/flux.py:301]: paraneters -> parameters

File: sherpa/astro/flux.py
Comment (wmclaugh) [inline sherpa/astro/flux.py:371]: For consistency, shouyld "hard" be 'hard'.  Looks like single quotes are used elsewhere...

File: sherpa/astro/flux.py
Comment (wmclaugh) [inline sherpa/astro/flux.py:497]: use -> used

File: sherpa/astro/ui/tests/test_astro_ui_fluxes.py
Comment (wmclaugh) [inline sherpa/astro/ui/tests/test_astro_ui_fluxes.py:830]: represention -> representation

File: sherpa/sim/sample.py
Comment (wmclaugh) [inline sherpa/sim/sample.py:520]: freedon -> freedom

File: sherpa/sim/sample.py
Comment (wmclaugh) [inline sherpa/sim/sample.py:730]: freedon -> freedom

**Comment by wmclaugh** [comment]
@DougBurke, this needs a rebase so that it can be merged.

File: sherpa/astro/flux.py
Comment (DougBurke) [reply sherpa/astro/flux.py]: Ta - I've made the doc changes.

**Comment by DougBurke** [comment]
@wmclaugh - let's hope for less fun with this rebase than my recent PRs!

**Comment by DougBurke** [comment]
I've had to restart the macOS test as it timed out (after a failed test, but in code not related to this PR so it may have failed for the same reason the whole test failed)

**Comment by DougBurke** [comment]
Yay it all passed.

**Pull request merged**
